In [1]:
import numpy as np
import cv2
from matplotlib import pyplot as plt

img1 = cv2.imread('dataset/0012_3_1.bmp',cv2.IMREAD_GRAYSCALE)        
img2 = cv2.imread('dataset/012_8_1.bmp',cv2.IMREAD_GRAYSCALE)           
img3 = cv2.imread('dataset/0012_3_2.bmp',cv2.IMREAD_GRAYSCALE)        
img4 = cv2.imread('dataset/0012_3_12.bmp',cv2.IMREAD_GRAYSCALE)# rotate of img1
img5 = cv2.imread('dataset/0012_3_13.bmp',cv2.IMREAD_GRAYSCALE)# rotate of img1
# Initiate ORB detector
orb = cv2.ORB_create()

# find the keypoints and descriptors with SIFT
kp1, des1 = orb.detectAndCompute(img1,None)
kp2, des2 = orb.detectAndCompute(img2,None)
kp3, des3 = orb.detectAndCompute(img3,None)
kp4, des4 = orb.detectAndCompute(img4,None)
kp5, des5 = orb.detectAndCompute(img5,None)

In [3]:
def get_ave_dist(matches):
    tmp = []
    k=200
    if(len(matches)>k):
        tmp[:k]=matches[:k]
    else:
        tmp = matches
        
    ave=0    
    for x in tmp:
        ave+=x.distance
    return ave/len(tmp)

In [4]:
# match 1 and 2

# create BFMatcher object
bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)

# Match descriptors.
matches = bf.match(des1,des2)

# Sort them in the order of their distance.
matches = sorted(matches, key = lambda x:x.distance)
get_ave_dist(matches),len(matches)

(71.015, 275)

In [5]:
# match 1 and 3

# create BFMatcher object
bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)

# Match descriptors.
matches = bf.match(des1,des3)

# Sort them in the order of their distance.
matches = sorted(matches, key = lambda x:x.distance)
get_ave_dist(matches),len(matches)

(67.59, 259)

In [6]:
# match 2 and 3

# create BFMatcher object
bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)

# Match descriptors.
matches = bf.match(des2,des3)

# Sort them in the order of their distance.
matches = sorted(matches, key = lambda x:x.distance)
get_ave_dist(matches),len(matches)

(72.02, 244)

In [7]:
# match 1 and 4

# create BFMatcher object
bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)

# Match descriptors.
matches = bf.match(des1,des4)

# Sort them in the order of their distance.
matches = sorted(matches, key = lambda x:x.distance)
get_ave_dist(matches),len(matches)

(3.5, 441)

In [8]:
# match 1 and 5

# create BFMatcher object
bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)

# Match descriptors.
matches = bf.match(des1,des5)

# Sort them in the order of their distance.
matches = sorted(matches, key = lambda x:x.distance)
get_ave_dist(matches),len(matches)

(3.605, 442)

In [9]:
# match 4 and 5

# create BFMatcher object
bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)

# Match descriptors.
matches = bf.match(des4,des5)

# Sort them in the order of their distance.
matches = sorted(matches, key = lambda x:x.distance)
get_ave_dist(matches),len(matches)

(0.0, 492)

# Summary

* img4,img5 are rotation of img1
* so, dist(img1,img4)=3.5, dist(img1,img5)=3.605, dist(img4,img5)=0.0